src/coreclr/jit/morph.cpp

```diff
@@ -8312,6 +8312,26 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
     call = fgMorphArgs(call);
     noway_assert(call->gtOper == GT_CALL);
 
+    // Try to replace CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE with a constant gc handle
+    // pointing to a frozen segment
+    if (!gtIsActiveCSE_Candidate(call) && gtIsTypeHandleToRuntimeTypeHelper(call))
+    {
+        GenTree*             argNode = call->AsCall()->gtArgs.GetArgByIndex(0)->GetNode();
+        CORINFO_CLASS_HANDLE hClass  = gtGetHelperArgClassHandle(argNode);
+        if (!opts.IsReadyToRun() && (hClass != NO_CLASS_HANDLE) && !gtIsActiveCSE_Candidate(argNode))
```

Comment:
What are all the places that make this incompatible with ReadyToRun compilation (NativeAOT also returns true for this).

I see the `CORINFO_RuntimeType` where JIT is now making assumptions about how RuntimeType looks like. But that's just to get a name. We could have a dedicated JitInterface API for that and keep it a handle.

Frozen RuntimeType instances can be supportable on NativeAOT side as well if RyuJIT treats RuntimeTypes like any other handles and doesn't do the CORINFO_RuntimeType with actual fields.

Reply:
`CORINFO_RuntimeType` is only used to get a string name for a type in JitDisasm comments - it's probably better to give up on it than introducing a new JIT-EE interface 😄

Reply:
Can you use `GenTreeIntCon::gtCompileTimeHandle` for this purpose? Initialize it to point to CORINFO_CLASS_HANDLE, etc. and then get the name from the CORINFO_CLASS_HANDLE using existing JIT/EE interface method?